nx/test/nx/vectorize_test.exs

```diff
@@ -512,4 +512,195 @@ defmodule Nx.VectorizeTest do
+               "\n"
+             ])
+  end
+
+  defn cond4(p1, c1, p2, c2, p3, c3, c4) do
+    cond do
+      p1 -> c1
+      p2 -> c2
+      p3 -> c3
+      true -> c4
+    end
+  end
+
+  test "1 vectorized pred in the beginning" do
+    assert cond4(Nx.vectorize(~V[0 1], :pred), 10, 0, 20, 0, 30, 40) ==
+             Nx.vectorize(~V[40 10], :pred)
+
+    assert cond4(Nx.vectorize(~V[0 0], :pred), 10, 1, 20, 0, 30, 40) ==
+             Nx.vectorize(~V[20 20], :pred)
+
+    assert cond4(Nx.vectorize(~V[0 0], :pred), 10, 0, 20, 1, 30, 40) ==
+             Nx.vectorize(~V[30 30], :pred)
+
+    assert cond4(Nx.vectorize(~V[0 0], :pred), 10, 0, 20, 0, 30, 40) ==
+             Nx.vectorize(~V[40 40], :pred)
+  end
+
+  test "1 vectorized pred in the second but not last position" do
+    assert cond4(0, 10, Nx.vectorize(~V[0 1], :pred), 20, 0, 30, 40) ==
+             Nx.vectorize(~V[40 20], :pred)
+
+    assert cond4(1, 10, Nx.vectorize(~V[0 1], :pred), 20, 0, 30, 40) ==
+             Nx.vectorize(~V[10 10], :pred)
+
+    assert cond4(0, 10, Nx.vectorize(~V[0 0], :pred), 20, 1, 30, 40) ==
+             Nx.vectorize(~V[30 30], :pred)
+
+    assert cond4(0, 10, Nx.vectorize(~V[0 0], :pred), 20, 0, 30, 40) ==
+             Nx.vectorize(~V[40 40], :pred)
+  end
+
+  test "1 vectorized pred in the last position" do
+    assert cond4(0, 10, 0, 20, Nx.vectorize(~V[0 1], :pred), 30, 40) ==
+             Nx.vectorize(~V[40 30], :pred)
+
+    assert cond4(1, 10, 0, 20, Nx.vectorize(~V[0 1], :pred), 30, 40) ==
+             Nx.vectorize(~V[10 10], :pred)
+
+    assert cond4(0, 10, 1, 20, Nx.vectorize(~V[0 1], :pred), 30, 40) ==
+             Nx.vectorize(~V[20 20], :pred)
+
+    assert cond4(0, 10, 0, 20, Nx.vectorize(~V[0 0], :pred), 30, 40) ==
+             Nx.vectorize(~V[40 40], :pred)
+  end
+
+  test "2 vectorized preds with different axes" do
+    assert cond4(
+             Nx.vectorize(~V[0 1 0], :pred1),
+             10,
+             Nx.vectorize(~V[1 0], :pred2),
+             20,
+             0,
+             30,
+             40
+           ) ==
+             Nx.vectorize(~M[
+              20 40
+              10 10
+              20 40
+            ], pred1: 3, pred2: 2)
+  end
+
+  test "2 vectorized preds with different axes + clauses that match either" do
+    [p1, p2] =
+      Nx.broadcast_vectors([Nx.vectorize(~V[0 1 0], :pred1), Nx.vectorize(~V[1 0], :pred2)])
+
+    cond4(
+      p1,
+      Nx.vectorize(~V[10 100], :pred2),
+      p2,
+      Nx.vectorize(~V[20 200 2000], :pred1),
+      0,
+      30,
+      40
+    )
+    |> IO.inspect()
+
+    # the results here differ, which maybe indicates that we do indeed need to broadcast vectorized_axes between vectorized predicates
+
+    assert cond4(
+             Nx.vectorize(~V[0 1 0], :pred1),
+             Nx.vectorize(~V[10 100], :pred2),
+             Nx.vectorize(~V[1 0], :pred2),
+             Nx.vectorize(~V[20 200 2000], :pred1),
```

Comment:
@josevalim I think this is the last point of contention between our solutions. pred1 and pred2 are indeed orthogonal, but I think that applying them independently might yield hard to predict results.

By doing the explicit broadcast between them, the result is easier to reason about, and we can do this broadcasting only between the already vectorized predicates.

Reply:
The result on the assertion below is the result given the explicit broadcasting.

Reply:
Ok, so after unwrapping this by hand, I decided that we need to at least reshape_vectors the preds together.
What's happening currently is that the first clause is vectorizing to [:pred1, :pred2] while the second clause is vectorizing to [:pred2, :pred1] and then being reshaped, so it yields unexpected results.

`reshape_vectors` takes care of this.